### ISSUE TITLE
Relationship update don't invoke updateRelationship policy method

### ISSUE THREAD

**Issue by bbprojectnet** [laravel-json-api/laravel#98]
When i try to update relationship by updating parent resource, like this:

```
		$response = $this->jsonApi()
			->includePaths('signatureAttachment')
			->withData([
				'type' => 'declarations',
				'id' => $declaration->getEncodedKey(),
				'relationships' => [
					'signatureAttachment' => [
						'data' => [
							'type' => 'attachments',
							'id' => $attachment->getEncodedKey(),
						],
					],
				],
			])
			->patch('/api/declarations/' . $declaration->getEncodedKey());
```
the `updateSignatureAttachment()` in `DeclarationPolicy` is not invoked, why?

**Comment by lindyhopchris** [comment]
This is intentional. The policy methods are there to authorizer the *controller actions*. They don't parse the JSON payload and authorizer individual parts of the JSON.

If you want to do this, you can either use the `authorize` method on the `Request` class to do your own authorization. Or you can write your own authorizer.

**Comment by bbprojectnet** [comment]
Hi,

Thanks for answer. Can you show my any use case of that behaviour? Because updating relation via /relationships/xxx or via updating resource itself it's the same action - has the same effect. But in one scenario, appropriate policy method is invoked, but in other no. It is inconsistent. Now I understand how it works under hood, but from developer side, it's very hard to guess that enabling resource attributes update, also enable a way to skip policy - it's sounds like security issue for me.

I thinks it should by written in documentation, that relationships policy methods are useless when you can update base resource it self.

**Comment by lindyhopchris** [comment]
Happy to discuss this a bit further.

How would you envisage it working? For example, if the client sends a relationship in the request to update a specific resource. Presumably you'd expect the authentication method for the relationship on the policy to be invoked - regardless of whether the provided value changes the relationship value? What happens if it would not change the value?

Another thing to discuss is that I have implemented everything to follow the pattern set by Laravel Nova (wherever possible). It's unclear to me in the Nova docs whether the policies are called in the way you are suggesting. Do you (or anyone else reading this) know for sure what the Nova behaviour is with relationship authorisation? Does it invoke the relationship authorisation method for each relationship in an update resource request? Having re-looked at the Nova docs, it's unclear to me.

**Comment by bbprojectnet** [comment]
> How would you envisage it working?

`UPDATE /resource/relationships/name` (whatever is changed or not) -> call updateName()
`UPDATE /resource` (without relationship section, whatever attributes are changed or not) -> call update()
`UPDATE /resource` (with relationship section, whatever is changed or not) -> call update() && updateName()

Two first is a current behaviour, only the last one is different.

> What happens if it would not change the value?

It should call policy, same as when you `UPDATE /resource/relationships/name` - even if you don't change, policy is called.

I never use Laravel Nova so it's hard to say :(

For now, it's very hard to stay secure with relationships, because stil exists a way to ommit policies - so they don't protect anything.

**Comment by lindyhopchris** [comment]
Yeah ok that's what I thought.

Next question... what to do with relationships that are read-only? In the way you've described, the relationship would be authorised when updating the whole resource, even if the value is not actually going to be filled because the developer has set it to read-only (or is not validating it, which is equivalent to ignoring it).

**Comment by lindyhopchris** [comment]
And another question.

At the moment when the resource is being viewed, we don't call all the `view*` relationship authorisation methods... Are you proposing we do? Which would then lead to the question, when viewing many of the resources, do we need to authorise the relationship for every resource that is going to be included in the response?!

As you can see, this is a rabbit-hole that you quickly disappear down.... and none of the applications I've designed and implemented have ever need this kind of authorisation logic.

**Comment by bbprojectnet** [comment]
> what to do with relationships that are read-only?

If relationships is read-only, you don't need to authorize it, because even if you include relationships section in request it should take no effect at all - like in read-only attributes.

> At the moment when the resource is being viewed, we don't call all the view* relationship authorisation methods... Are you proposing we do? 

No, if you just view resource you just view resource ;) - in that case we don't even have any relationship name to call policy ;)

You should call view*() when GET /resource/relationships/name, or include=...

> and none of the applications I've designed and implemented have ever need this kind of authorisation logic.

So how you protect your relationships? Let's suppose we have some resource with 2 relationships, first can be changes by anyone, the second, only by the author of resource. In current beahaviour anyone can change first and second by just by updating resource it self, because policy only protect /resource/relationships/xxx URL.

**Comment by lindyhopchris** [comment]
Have pondered this a bit more.

Basically, you're trying to do **field-level authorization**... i.e. access rights can vary at a resource field (attribute/relationship) level. The authorizer in this package does **resource-level** authorization.

Having thought about this more, my preference is for this package to continue to support **resource-level** authorization. It does support *field-level* - you'd just need to ship your own authorizer.

The reason for keeping it at the *resource-level* is it is the most straight-forward strategy and will work with the majority of use-cases. In fact, I would always design an API that way personally, as I think it is the easiest strategy to explain in API documentation. If I did need to protect a specific attribute or relationship, I'd use the read-only feature to make an attribute/relationship read-only if a user didn't have a permission. Alternatively, for complex model authorization I tend to split the model into several different resources (using our proxy-model feature) to control access that way; or I split them into separate APIs. For example the main app I use this package in has one API for a user to access their own data, and another API for an event organiser to access data as an event organiser. This keeps the authorization at the *resource-level* per API.

None of this is saying you shouldn't do it the way you're wanting to.... *field-level* authorization is totally a choice the designer of an API can take. However, having thought about it there's way too many application-specific choices as to how to do this. Also, if you want to protect it per-model that is loaded (either when listing resources, or when eager loading to-many relationships) you'd actually need to implement it lower down the stack as you'd be reliant on the data loading *first*, then authorizing *second*.

In summary, this package will continue supporting *resource-level* authorization. *Field-level* authorization is supported by you being able to write and use your own authorizers.

This is now explained in the docs here:
https://laraveljsonapi.io/docs/1.0/requests/authorization.html#relationship-authorization
